pkg/client/unversioned/clientcmd/merged_client_builder.go

```diff
@@ -115,10 +115,10 @@ func (config *DeferredLoadingClientConfig) ClientConfig() (*restclient.Config, e
 		// TODO: this shouldn't be a global - the client config rules should be
 		//   handling this.
 		defaultConfig, defErr := DefaultClientConfig.ClientConfig()
-		if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
+		if defErr != nil && !IsConfigurationInvalid(defErr) {
 			return mergedConfig, nil
```

Comment:
You're right, an empty default config is equivalent to an invalid default config.  And that really points out that Chao was right - we shouldn't even be checking errors here, and we shouldn't be in this case block at all when we have an empty default config (it's more confusion than it's worth).

I'd prefer not to revert to the global for when the factory is loaded - it would be changing a global after initialization which means we have to deal with thread safety issues, and we can't default it globally because that breaks hyperkube (it reverts to the insecure behavior for proxy since ).  The default kubectl settings don't mandate in cluster config unless localhost:8080 is defined to be the cluster, which is impossible for in-cluster config today and likely will remain that way forever.  I'd prefer not to broaden ICC anymore than we have to, especially since it's effectively a MITM to default to http://localhost:8080 for master communications.

Looking at BuildFromKubeconfigGetter, it's using Load() which returns an empty data on an empty secret.  Can you point me to the test case that is failing (since Gubernator seems to be down)?  Do you have a valid secret at that point?  But it also looks like you are doing some sort of InClusterConfig detection up front (https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/cluster_util.go#L105) which probably should be part of clientcmd in the future.